File: amath 483/homework3/intersections.f90
! $UWHPSC/codes/fortran/newton/intersections.f90

program intersections

    use newton, only: solve
    use functions, only: g1g2, g1g2p
    
    implicit none
    real(kind=8) :: x, x0, fx
    real(kind=8) :: x0vals(4)
    integer :: iters, itest
    logical :: debug

    debug = .false.
    x0vals = (/-2.2d0,-1.6d0,-0.77d0,1.43d0/)
    
    do itest = 1,4
	x0 = x0vals(itest)

        print 11, x0
11      format('With the initial guess x0 = ', e22.15)

 	call solve(g1g2, g1g2p, x0, x, iters, debug)

        print 12, x, iters
12      format('	 solver returns x = ', e22.15, ' after', i3, ' iterations')

        fx = g1g2(x)

        if (abs(fx) > 1d-14) then
            print 13, x
13          format('*** Unexpected result: x = ', e22.15)
            endif
		print *, ' '  ! blank line
        enddo

end program intersections

File: amath 483/homework3/intersections.py
from pylab import *
from newton import solve

def g1(x):
    return x * cos(pi*x)

def g2(x):
    return 1 - .6*x**2

def g1g2(x):
    f = g1(x) - g2(x)
    fp = cos(pi*x) - x*pi*sin(pi*x) +  1.2*x
    return f, fp

##calculating the 4 zeros from the 4 initial guesses.
x_guesses = [-2.2,-1.6,-0.77,1.43]
zeros = []
for x0 in x_guesses:
    x, i = solve(g1g2, x0, False)
    zeros.append(x)
    print "With initial guess x0 = %22.15e," %x0
    print "      solve returns x = %22.15e after %i iterations" %(x,i)
    print

##plot the results
x = linspace(-5,5,500)
figure(1)
clf()
plot(x,g1(x))
plot(x,g2(x))
for i in range(4):
    plot(zeros[i],g1(zeros[i]),'ko')
savefig('intersections.png')



